File: 3418. maximum-amount-of-money-robot-can-earn.py
class Solution(object):
    def maximumAmount(self, coins):
        n = len(coins)
        m = len(coins[0])
        dp = [[[-10**9] * 3 for _ in range(m)] for _ in range(n)]

        dp[0][0][1] = dp[0][0][2] = 0
        dp[0][0][0] = coins[0][0]

        for i in range(n):
            for j in range(m):
                for k in range(3):
                    if i:
                        dp[i][j][k] = max(dp[i][j][k], dp[i - 1][j][k] + coins[i][j])
                    if i and k:
                        dp[i][j][k] = max(dp[i][j][k], dp[i - 1][j][k - 1])
                    if j:
                        dp[i][j][k] = max(dp[i][j][k], dp[i][j - 1][k] + coins[i][j])
                    if j and k:
                        dp[i][j][k] = max(dp[i][j][k], dp[i][j - 1][k - 1])

        return max(dp[n - 1][m - 1])

if __name__ == "__main__":
    sol = Solution()
    
    # Example 1
    coins1 = [[0,1,-1],[1,-2,3],[2,-3,4]]
    output1 = sol.maximumAmount(coins1)
    print(f"Example 1: coins=[[0,1,-1],[1,-2,3],[2,-3,4]]")
    print(f"Output: {output1}")
    print(f"Expected: 8")
    print()
    
    # Example 2
    coins2 = [[10,10,10],[10,10,10]]
    output2 = sol.maximumAmount(coins2)
    print(f"Example 2: coins=[[10,10,10],[10,10,10]]")
    print(f"Output: {output2}")
    print(f"Expected: 40")

File: 3418. maximum-amount-of-money-robot-can-earn.java
import java.util.*;

class Solution {
    public int maximumAmount(int[][] coins) {
        int n = coins.length;
        int m = coins[0].length;
        int[][][] dp = new int[n][m][3];

        for (int[][] row : dp) {
            for (int[] col : row) {
                java.util.Arrays.fill(col, (int)-1e9);
            }
        }

        dp[0][0][1] = 0;
        dp[0][0][2] = 0;
        dp[0][0][0] = coins[0][0];

        for (int i = 0; i < n; i++)
            for (int j = 0; j < m; j++)
                for (int k = 0; k < 3; k++) {
                    if (i > 0) dp[i][j][k] = Math.max(dp[i][j][k], dp[i - 1][j][k] + coins[i][j]);
                    if (i > 0 && k > 0) dp[i][j][k] = Math.max(dp[i][j][k], dp[i - 1][j][k - 1]);
                    if (j > 0) dp[i][j][k] = Math.max(dp[i][j][k], dp[i][j - 1][k] + coins[i][j]);
                    if (j > 0 && k > 0) dp[i][j][k] = Math.max(dp[i][j][k], dp[i][j - 1][k - 1]);
                }

        return Math.max(dp[n - 1][m - 1][0], Math.max(dp[n - 1][m - 1][1], dp[n - 1][m - 1][2]));
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Example 1
        int[][] coins1 = {{0,1,-1},{1,-2,3},{2,-3,4}};
        int output1 = sol.maximumAmount(coins1);
        System.out.println("Example 1: coins=[[0,1,-1],[1,-2,3],[2,-3,4]]");
        System.out.println("Output: " + output1);
        System.out.println("Expected: 8");
        System.out.println();
        
        // Example 2
        int[][] coins2 = {{10,10,10},{10,10,10}};
        int output2 = sol.maximumAmount(coins2);
        System.out.println("Example 2: coins=[[10,10,10],[10,10,10]]");
        System.out.println("Output: " + output2);
        System.out.println("Expected: 40");
    }
}